app/models/user.rb

```diff
@@ -163,31 +163,43 @@ def course_student?
   end
 
   def nonvisitor?(course)
-    role(course) != CoursesUsers::Roles::VISITOR_ROLE
+    return true if admin?
+    roles = course_roles(course)
+    roles.size > 1 || (roles.size == 1 && roles.first != CoursesUsers::Roles::VISITOR_ROLE)
+  end
+
+  # returns an array of roles a user has in a given course
+  def course_roles(course)
+    user_course_roles = course.courses_users.where(user_id: id).order('role DESC').pluck(:role)
+    return user_course_roles
```

Comment:
Better to make this a one line method without assigning it to a variable or using explicit return.